pelican/log.py

```diff
@@ -1,142 +1,446 @@
     )
+    cfg.update(settings.copy())
+
+    # validate & transform types/values: treat ``None`` as logging.NOTSET
+    if not isinstance(cfg.get('LOG_ONCE_LEVEL') or logging.NOTSET, int):
+        cfg['LOG_ONCE_LEVEL'] = severity_from_name(cfg['LOG_ONCE_LEVEL'])
 
-    logger = logging.getLogger(name)
+    if not isinstance(cfg.get('LOG_LOGGER'), logging.Logger):
+        cfg['LOG_LOGGER'] = logging.getLogger(cfg.get('LOG_LOGGER'))
 
-    if level:
-        logger.setLevel(level)
-    if logs_dedup_min_level:
-        LimitFilter.LOGS_DEDUP_MIN_LEVEL = logs_dedup_min_level
+    # == FILTERING:
+    logger = set_filters(
+        logger=cfg.get('LOG_LOGGER'),
+        limit_threshold=cfg.get('LOG_LIMIT_THRESHOLD'),
+        once_lvl=cfg.get('LOG_ONCE_LEVEL'),
+        blacklist=cfg.get('LOG_FILTER')
+    )
 
+    # == HANDLING:
+    if cfg.get('LOG_FATAL'):
+        hnd = AbortHandler(cfg['LOG_FATAL'])
 
-def log_warnings():
-    import warnings
-    logging.captureWarnings(True)
-    warnings.simplefilter("default", DeprecationWarning)
-    init(logging.DEBUG, name='py.warnings')
+        # Add to root logger so all other handlers can write/emit first.
+        root = logging.getLogger()
+        if logger is not root:
+            # handler only acts on records from specified logger (& children).
+            hnd.addFilter(logging.Filter(logger.name))
+        root.addHandler(hnd)
 
+    return logger
 
-if __name__ == '__main__':
-    init(level=logging.DEBUG, name=__name__)
 
-    root_logger = logging.getLogger(__name__)
-    root_logger.debug('debug')
-    root_logger.info('info')
-    root_logger.warning('warning')
-    root_logger.error('error')
-    root_logger.critical('critical')
+def set_filters(
+    logger=None,  # logger of interest
+    limit_threshold=None,  # LimitFilter: when to start using limit_msg
+    once_lvl=None,  # OnceFilter: which severity to dedupe msgs at
+    blacklist=None  # BlacklistFilter:
+):
+    """
+    Apply filtering on the logger (Default: ``PKG_LOGGER``).
+
+    Up to three filters are (optionally) applied:
+        ``LimitFilter``:
+            Modifies log records using `limit_msg` in the `extras` dict. The
+            `limit_threshold` specifies how many records using the same
+            limit_msg are emitted before converting the message.  See
+            ``LimitFilter`` for an example.
+        ``OnceFilter``:
+            De-duplicates messages of identical severity & content. Only
+            affects messages of once_lvl and higher.
+        ``BlacklistFilter``:
+            Prevents log records of (severity, msg) from being emitted.
+
+    Returns
+    -------
+    The logger object the filters were applied to.
+    """
+    if logger is None:
+        logger = PKG_LOGGER
+
+    if blacklist:
+        try:
+            flt = BlacklistFilter.get_filters(logger)[0]
+        except IndexError:
+            flt = BlacklistFilter()
+            logger.addFilter(flt)
```

Comment:
Also, might want to rename `BlacklistFilter` :). 
`IgnoreFilter`? 
`SuppressFilter`?

Reply:
This can be renamed, yes; I was ignorant of the PC this term violates.  I'm leaning towards ``LogCensor``, ``ExclusionFilter``, or ``RecordBan``.  

As per the ``get_filters()`` method: pelican's current implementation requires a singleton to allow a list of banned records to be updated (or replaced).  This interface gives the caller access to any such filters on a specific logger, so they may be updated (or replaced) *without* needing to keep a reference to the filter directly.  All that's needed here is to provide the targeted logger, which can be retrieved from any scope; again *without* needing to keep a reference to any filters that may or may not be in place.

Reply:
I am fine with `get_filters` method. I'm talking about the logic within `set_filters`. Specifically the reason to check if there was a already a filter on a logger. As far as I can see, `set_filters` *is* intended interface put filters to a logger.

So, this check means either `set_filters` can be run multiple times on the same logger or something besides `set_filters` will assign filters beforehand. I don't see why any of them would happen, so I'm asking will these ever happen?

Reply:
If a config's LOG_FILTER value changes, I'd expect the logging behavior to update when the config is reloaded.

Though looking at this again now, I guess I'd potentially need to reset/replace other filters as well, especially as I'm hoping to eventually support more logging configuration options through the configuration file.

I'll need to take another look at how config reloading is implemented.

Reply:
> So, this check means either set_filters can be run multiple times on the same logger or something besides set_filters will assign filters beforehand. I don't see why any of them would happen, so I'm asking will these ever happen?

I should also say that I'm writing the logging configuration with the intention of allowing a user/developer to customize the logger configuration outside of the pelican library (in some wrapper, or even in ``pelicanconf.py``).  

That said, I'm now revisiting this and trying to wrap my head around how ``configure()`` & ``set_filters()`` should behave on pre-configured loggers / pre-existing filters...